acl_loader/main.py

```diff
@@ -228,25 +265,72 @@ def convert_action(self, table_name, rule_idx, rule):
 
         if rule.actions.config.forwarding_action == "ACCEPT":
             if self.is_table_control_plane(table_name):
-                rule_props["PACKET_ACTION"] = "ACCEPT"
+                rule_props[AclAction.PACKET] = PacketAction.ACCEPT
             elif self.is_table_mirror(table_name):
                 session_name = self.get_session_name()
                 if not session_name:
                     raise AclLoaderException("Mirroring session does not exist")
 
-                rule_props["MIRROR_ACTION"] = session_name
+                if self.mirror_stage == Stage.INGRESS:
+                    mirror_action = AclAction.MIRROR_INGRESS
+                elif self.mirror_stage == Stage.EGRESS:
+                    mirror_action = AclAction.MIRROR_EGRESS
+                else:
+                    raise AclLoaderException("Invalid mirror stage passed {}".format(self.mirror_stage))
+
+                rule_props[mirror_action] = session_name
             else:
-                rule_props["PACKET_ACTION"] = "FORWARD"
+                rule_props[AclAction.PACKET] = PacketAction.FORWARD
         elif rule.actions.config.forwarding_action == "DROP":
-            rule_props["PACKET_ACTION"] = "DROP"
+            rule_props[AclAction.PACKET] = PacketAction.DROP
         elif rule.actions.config.forwarding_action == "REJECT":
-            rule_props["PACKET_ACTION"] = "DROP"
+            rule_props[AclAction.PACKET] = PacketAction.DROP
         else:
-            raise AclLoaderException("Unknown rule action %s in table %s, rule %d" % (
+            raise AclLoaderException("Unknown rule action {} in table {}, rule {}".format(
+                rule.actions.config.forwarding_action, table_name, rule_idx))
+
+        if not self.validate_actions(table_name, rule_props):
+            raise AclLoaderException("Rule action {} is not supported in table {}, rule {}".format(
                 rule.actions.config.forwarding_action, table_name, rule_idx))
 
         return rule_props
 
+    def validate_actions(self, table_name, action_props):
+        if self.is_table_control_plane(table_name):
+            return True
+
+        action_count = len(action_props)
+
+        if table_name not in self.tables_db_info:
+            raise AclLoaderException("Table {} does not exist".format(table_name))
+
+        stage = self.tables_db_info[table_name].get("stage", Stage.INGRESS)
+        capability = self.statedb.get_all(self.statedb.STATE_DB, "{}|switch".format(self.SWITCH_CAPABILITY_TABLE))
+        for action_key in dict(action_props):
+            key = "{}|{}".format(self.ACL_ACTIONS_CAPABILITY_FIELD, stage.upper())
+            if key not in capability:
+                del action_props[action_key]
```

Comment:
why not raise immediately?

Reply:
this method ```validate_actions``` can validate several actions in one call. This is done to allow later check if some combination of actions in one rule is allowed